pkg/comp-functions/functions/spks/spksredis/tls.go

```diff
@@ -0,0 +1,210 @@
+		}
+
+		err = common.SetNestedObjectValue(values, []string{"replica", "replicaCount"}, 0)
+		if err != nil {
+			return err
+		}
+
+		rawValues, err := json.Marshal(values)
+		if err != nil {
+			return err
+		}
+
+		desRelease.Spec.ForProvider.Values.Raw = rawValues
+
+		err = updateScaleTimestamp(svc, comp)
+		if err != nil {
+			return fmt.Errorf("cannot update composite status: %w", err)
+		}
+
+	}
+
+	err = svc.SetDesiredComposedResourceWithName(desRelease, "redis-chart")
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getRedisReleaseTLSStatus(release *xhelm.Release, svc *runtime.ServiceRuntime) bool {
+
+	rawValues := release.Spec.ForProvider.Values
+
+	values, err := convertValuesToMap(rawValues.Raw)
+	if err != nil {
+		svc.Log.Info("cannot marshal release tls status: " + err.Error())
+		return false
+	}
+
+	enabled, _, err := unstructured.NestedBool(values, "tls", "enabled")
+	if err != nil {
+		svc.Log.Info("cannot get tls status from values: " + err.Error())
+		return false
+	}
+
+	return enabled
+}
+
+func convertValuesToMap(rawValues []byte) (map[string]any, error) {
+	values := map[string]any{}
+	err := json.Unmarshal(rawValues, &values)
+	if err != nil {
+		return nil, err
+	}
+
+	return values, nil
+}
+
+func addSTSObserver(comp *spksv1alpha1.CompositeRedisInstance, svc *runtime.ServiceRuntime) error {
+	sts := &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "redis-node",
+			Namespace: comp.GetName(),
+		},
+	}
+
+	err := svc.SetDesiredKubeObject(sts, comp.GetName()+"-sts-observer", runtime.KubeOptionObserve)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// isSTSScaling will return true, if the release it set to 0 replicas and
+// the sts hasn't fully scaled down yet.
+// This is necessary to ensure that we don't have mixed tls/non-tls members in the cluster
+// at the same time.
+func isSTSScaling(obsRelease *xhelm.Release, comp *spksv1alpha1.CompositeRedisInstance, svc *runtime.ServiceRuntime) bool {
+	sts := &appsv1.StatefulSet{}
+
+	err := svc.GetObservedKubeObject(sts, comp.GetName()+"-sts-observer")
+	if err != nil {
+		svc.Log.Error(err, "cannot get sts observer")
+		return false
+	}
+
+	obsValues, err := convertValuesToMap(obsRelease.Spec.ForProvider.Values.Raw)
+	if err != nil {
+		svc.Log.Error(err, "cannot convert values to map")
+		return false
+	}
+
+	count, _, err := unstructured.NestedFloat64(obsValues, "replica", "replicaCount")
+	if err != nil {
+		svc.Log.Error(err, "cannot get replica.replicaCount")
+		return false
+	}
+
+	if count == 0 {
+		return !(sts.Status.Replicas == 0)
+	}
+
+	return false
+}
+
+// updateScaleTimestamp will update a field in the status called `scaleTimeStamp`.
+// This will instantly trigger a new reconcile.
+// Making the switch from TLS to non-TLS much faster on SPK. Because SPKS has a
+// default reconcile period of 10 minutes.
+func updateScaleTimestamp(svc *runtime.ServiceRuntime, comp *spksv1alpha1.CompositeRedisInstance) error {
```

Comment:
I hope this will not interfere in heavy used clusters such as prod. With how many instances have you tested this feature? (not just redis 7)

Reply:
It's currently only implement in Redis 7 and it will only trigger if you switch from tls to no-tls or vice-versa. It won't trigger on newly provisioned instances either.